### PR TITLE
Use XDG standard instead of writing to home directory

### PR DESCRIPTION
**Pull request by mstruebing**
The `now-cli` places its config inside the `~` directory. This is a bad pattern as there is a standard which describes where specific types of files needed by programs should be placed.

This PR introduces that `now-cli` now uses the correct path (either `XDG_DATA_HOME/now` or `~/.local/share/now` as fallback)

### CODE REVIEW

**Comment by juliangruber** [comment]
It looks like this is a *nix only convention. I'm +-0 on updating this (`~/.now` on Mac is a lot simpler), and if we were to merge this, we'd need a fallback strategy so users don't have to log in again when they already have their config in the old location.

**Comment by mstruebing** [comment]
Hey I've added the ability to move the old config into it's new location.
Now the integration tests are failing on `scale the alias`, this seems strange to me as I wouldn't think my change affects this one.
Can somebody help me out?

**Comment by mstruebing** [comment]
Any feedback on this? cc @juliangruber

**Comment by juliangruber** [comment]
@mstruebing CI passes, I just had to restart it.

I'm curious how many users are aware of the location of `~/.now`, if no one knows then this should be an easy change to roll out. I'll ask what the team thinks about this.

For extra clarity, this change is important to you because Now CLI currently places files in a non-standard directory, while for Linux a well accepted standard exists, right?

**Comment by mstruebing** [comment]
Basically because of this: https://0x46.net/thoughts/2019/02/01/dotfile-madness/

The home directory is cluttered with dotfiles from every program with different things like cache, configs etc. whereas it would be a good idea to store it in a place which is specified for it.  
I think this standard is well accepted by all unixy-like environments. If you have a Mac you could just have a look at what's inside `~/.local/share/` (or whatever your `XDG_DATA_HOME`-env variable is set to).

**Comment by rivy** [comment]
Just to chime in here on portability, the XDG convention isn't necessarily a *nix-only convention. Obviously, as it originated in a *nix distribution, *nix use is more common. But, as it only involves setting some environment variables for use, it can be useful within any environment (*nix, Mac, PC, ...). And it is used by many portable programs across those environments.

**Comment by mstruebing** [comment]
Any news what the team thinks @juliangruber ?

**Comment by juliangruber** [comment]
@mstruebing I'm not responsible for this any more, cc @leo

**Comment by TooTallNate** [comment]
Related to https://github.com/zeit/now-cli/issues/1276.

**Comment by mstruebing** [comment]
What's going on here?
It's a bit disappointing to do a PR and doesn't a proper review for half a year, even I understand that all of you are busy and that maybe isn't on your priority list. /cc @leo @TooTallNate

**Review by leo — CHANGES_REQUESTED** [review]
Thank you for this wonderful PR!

Is it possible to only consider the standard when creating **new** configs? Like this, we could drastically decrease the chance for errors and ship the PR immediately.

**Comment by mstruebing** [comment]
Thanks for your feedback.
This is currently the case. As you can see in https://github.com/zeit/now-cli/pull/1909/files#diff-9f4ea4db912432cb805e4eb013ef3759R46 
it checks if a config in the old place exists and if so moves it to the new location. 
I think this is the easiest way for the user, or do you mean to don't touch the old config and leave it where it is and simply return the old location when it exists?

File: src/util/config/global-path.ts [hunk 1]
```diff
@@ -17,7 +38,16 @@ const getNowDir = () => {
   const customPath = args['global-config'];
 
   if (!customPath) {
-    return path.join(homedir(), '.now');
+    const nowDir = path.join(
+      process.env.XDG_DATA_HOME || path.join(homedir(), '.local', 'share'),
```
Comment (TooTallNate) [inline src/util/config/global-path.ts]: How does this standard affect Windows? Or if it doesn't apply, is there a more standardized place for us to store these files on Windows? Perhaps `$HOME/AppData/co.zeit.now/Credentials`?

**Comment by TooTallNate** [comment]
Right, I think @leo is saying only in the logged out case would these files be created. If there's existing credentials in the old location than just _use_ those and don't bother with migrating them implicitly.

Also, would you mind updating with the latest `canary` branch? As it seems to have merge conflicts.

I am 👍 for this change btw, and apologize for the lack of response in a timely fashion.

File: src/util/config/global-path.ts [hunk 2]
```diff
@@ -17,7 +38,16 @@ const getNowDir = () => {
...
+      'now'
+    );
+
+    if (oldConfigExists(oldConfigPath)) {
+      mvOldConfig(oldConfigPath, nowDir);
```
Comment (TooTallNate) [inline src/util/config/global-path.ts]: Missing `await` here. But aside from that I think we don't want to move the old config if it exists, but rather just use it as-is.

File: src/util/config/global-path.ts (hunk 1 above)
Comment (rivy) [reply src/util/config/global-path.ts]: You might look at [xdg-app-paths](https://www.npmjs.com/package/xdg-app-paths) / [xdg-portable](https://www.npmjs.com/package/xdg-portable) (both written by me) which allows use of XDG on any platform with platform-compatible defaults. Specifically, [xdg-portable](https://www.npmjs.com/package/xdg-portable) has and discusses reasonable platform-dependent defaults.

File: src/util/config/global-path.ts (hunk 2 above)
Comment (mstruebing) [reply src/util/config/global-path.ts]: It's removed :)

Comment (mstruebing) [reply src/util/config/global-path.ts]: And I simply use the old config now.

File: src/util/config/global-path.ts (hunk 1 above)
Comment (mstruebing) [reply src/util/config/global-path.ts]: I'm using `xdg-portable` now and if nothing can be found the old location is used (which should not happen on normal systems I think)

**Comment by rivy** [comment]
For config, using [`xdg-app-paths`]() and code something like ...

```javascript
import * as locatePath from 'locate-path';
import * as appPaths from 'xdg-app-paths';
// config file
// * search for config file within user preferred directory list and prior default; otherwise, use preferred directory
const configFileName = '.now';
const possibleConfigPaths = appPaths.configDirs()
    .concat(path.join(os.homedir(), '.config'))
    .map(v => path.join(v, configFileName));
const configPath = locatePath.sync(possibleConfigPaths) || possibleConfigPaths[0];
// mkdirp.sync(path.dirname(configPath), 0o700);
```

might be a better algorithm.

**Comment by mstruebing** [comment]
Thanks, @rivy, could you maybe explain a bit what's not so cool about my code? Would be willing to know :)

**Comment by rivy** [comment]
@mstruebing , happy to critique 😄 

- I think you really want 'config' instead of 'data' for the config file (see https://www.npmjs.com/package/xdg-portable#methods for some descriptions of where different data types might best be placed).
- 'xdg-portable' already tests/uses the corresponding environment variable, so you don't need to separately test XDG_CONFIG_HOME (or XDG_DATA_HOME).
- Both `configDirs()` and `dataDirs()` return a priority list of possible config/data locations respectively which *includes* the XDG_CONFIG_HOME or XDG_DATA_HOME location as the first entry. So, the single `xdg.configDirs()` will get you what you need.

[`xdg-app-paths`](https://www.npmjs.com/package/xdg-app-paths) goes one step further than `xdg-portable`. `xdg-app-paths` abstracts out some platform-differences giving you locations which the application can have for it's exclusive use (ie, avoiding naming collisions with other apps). You can also access the "communal" directory by manipulating the `isolated` function option. There's some explanation in the `xdg-app-paths` [discussion section](https://www.npmjs.com/package/xdg-app-paths#discussion) which goes into some platform-dependent details.

The example code I supplied above looks for the *.now* config file in the "standard configuration locations", including XDG settings, if present, plus the original location, using any found file, dropping back to the first priority location, if no actual *.now* file is found.

Not really knowing the details bout the current configuration file setup (and naming), I didn't fully flesh out the example. You might really want something more like ...
``` javascript
import * as locatePath from 'locate-path';
import * as appPaths from 'xdg-app-paths';
// config file
// * search for config file within user preferred directory list and prior default; otherwise, use preferred directory
const configFileName = 'now';
const possibleConfigPaths = appPaths.configDirs()
    .map(v => path.join(v, configFileName))
    .concat(path.join(os.homedir(), '.now'));
const configPath = locatePath.sync(possibleConfigPaths) || possibleConfigPaths[0];
mkdirp.sync(path.dirname(configPath), 0o700);
```

As the configuration files should all be within a communal or application-specific subdirectory when using the XDG specification, you probably don't want to hide the config file hidden by using the leading '.'.

The `concat` when building the `possibleConfigPaths` array are just the possible location(s) of the config file from the current setup.

The code uses the first location of the config file found, with fallback to the user-determined first priority location. By doing it this way, any current config file will be used and the user can move it between compatible locations.

`appPaths.$name()` could also be used in place of `configFileName`.

Notably, `xdg-app-paths` tries to determine the application name from the script name. If it's named something like now-cli.js, you may need to specify the application name when initializing it (eg, `const appPaths = require('xdg-app-paths')('now')` or `import * as xdgAppPaths from 'xdg-app-paths'; const appPaths = xdgAppPaths('now');`).

I hope that's helpful.

**Comment by mstruebing** [comment]
Thanks for your detailed answer.

I think `data` is more what is stored in the config file to be honest.
There is an `auth.json` which contains an auth token and a `config.json` which contains a team-id (if joined any team).  That doesn't really looks like configuration for me until now, but maybe this will change in the future.

I only hide the directory in case it is the legacy one in the  home directory.

**Comment by leo** [comment]
We're code freezing today, because @styfle is creating a monorepo with `zeit/now-cli`, `zeit/now-builders` and `zeit/now-client` here.

After that, we can merge (assuming that all the other concerns in the PR are covered).

**Comment by mstruebing** [comment]
Thank you, I will do it this week :)

**Comment by mstruebing** [comment]
Sorry, I've messed up a bit with rebasing, so this is gotten into a new PR: #2877

**Pull request closed without merging**